src/resolve/resolvectl.c

```diff
@@ -108,51 +108,58 @@ static int interface_info_compare(const InterfaceInfo *a, const InterfaceInfo *b
+int ifname_mangle_full(const char *s, bool drop_protocol_specifier) {
+        _cleanup_free_ char *buf = NULL;
         int ifi;
 
         assert(s);
 
-        iface = strdup(s);
-        if (!iface)
-                return log_oom();
+        ifi = rtnl_resolve_interface(NULL, s);
+        if (ifi < 0 && drop_protocol_specifier) {
+                /* When invoked as resolvconf, drop the protocol specifier(s) at the end. */
+                for (;;) {
+                        char *dot;
+
+                        if (buf) {
+                                dot = strrchr(buf, '.');
+                                if (!dot)
+                                        break;
+
+                                *dot = '\0';
+                        } else {
+                                dot = strrchr(s, '.');
+                                if (!dot)
+                                        break;
+
+                                buf = strndup(s, dot - s);
+                                if (!buf)
+                                        return log_oom();
+                        }
 
-        ifi = rtnl_resolve_interface(NULL, iface);
-        if (ifi < 0) {
-                if (ifi == -ENODEV && arg_ifindex_permissive) {
-                        log_debug("Interface '%s' not found, but -f specified, ignoring.", iface);
-                        return 0; /* done */
+                        log_debug("Ignoring protocol specifier '%s'.", dot + 1);
+                        ifi = rtnl_resolve_interface(NULL, buf);
+                        if (ifi >= 0)
+                                break;
                 }
+        }
 
-                return log_error_errno(ifi, "Failed to resolve interface \"%s\": %m", iface);
+        if (ifi == -ENODEV && arg_ifindex_permissive) {
+                log_debug("Interface '%s' not found, but -f specified, ignoring.", s);
+                return 0; /* done */
         }
 
+        if (ifi < 0)
+                return log_error_errno(ifi, "Failed to resolve interface \"%s\": %m", s);
+
         if (arg_ifindex > 0 && arg_ifindex != ifi)
                 return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Specified multiple different interfaces. Refusing.");
 
         arg_ifindex = ifi;
-        free_and_replace(arg_ifname, iface);
+        if (!buf)
+                return free_and_strdup_warn(&arg_ifname, s);
 
-        return 1;
-}
-
-int ifname_resolvconf_mangle(const char *s) {
-        const char *dot;
-
-        assert(s);
-
-        dot = strrchr(s, '.');
-        if (dot) {
-                _cleanup_free_ char *iface = NULL;
-
-                log_debug("Ignoring protocol specifier '%s'.", dot + 1);
-                iface = strndup(s, dot - s);
-                if (!iface)
-                        return log_oom();
-                return ifname_mangle(iface);
-        } else
-                return ifname_mangle(s);
+        free_and_replace(arg_ifname, buf);
+        return 1; /* found */
 }
```

Comment:
so i really dislike the ambiguity this introduces, i.e. you never know what "foo.bar" means: interface "foo" with protocol "bar", or interface "foo.bar" with no protocol?

While I think your patch generally makes sense I think we should at the very least warn about this ambiguity, and in selected cases even refuse. Specifically:

1. If there's practical ambiguity (and not just theoretical ambiguity) we should refuse. i.e. of the user passes "foo.bar" as interface name and there is both an interface "foo.bar" *and* one called "foo" existing, then we really don't know what to make of this.

2. if the user passes in an interface name containing a dot and it exists we should use it, but warn.

(conversely, we should not warn for interfaces that have no dot in them, because then things are clear. And we should not warn about interfaces that have a dot in them, and there's none by that full name, but once we chop off one dot it does exist).